src/Database/LSMTree/Internal/MergingTree.hs

```diff
@@ -178,10 +180,12 @@ newPendingUnionMerge ::
   => [Ref (MergingTree m h)]
   -> m (Ref (MergingTree m h))
 newPendingUnionMerge mts = do
-    mts' <- mapM dupRef =<< filterM (fmap not . isStructurallyEmpty) mts
-    case mts' of
-      [mt] -> return mt
-      _    -> newMergeTree (PendingTreeMerge (PendingUnionMerge mts'))
+    mts' <- V.mapM dupRef
+        =<< V.filterM (fmap not . isStructurallyEmpty) (V.fromList mts)
+    case V.uncons mts' of
+      Just (mt, mts'') | V.null mts''
+        -> return mt
+      _ -> newMergeTree (PendingTreeMerge (PendingUnionMerge mts'))
```

Comment:
This can leak references. If a `dupRef` fails, we won't release any of the references that were successfully dupped before that

Reply:
`dupRef` cannot fail if we have async exceptions masked, which we must do since `newPendingUnion` creates a new resource.

Reply:
I've thought before that what would be nice, is runtime or type level assertions that check that async exceptions really are masked where we expect them to be (and perhaps help us avoid repeated masking where it's not necessary).

Reply:
BTW, this code can get async exceptions because `isStructurallyEmpty` has an interruptible `readMVar` operation inside. So even if `dupRef` can't fail, then `isStructurallyEmpty` can

Reply:
> BTW, this code can get async exceptions because isStructurallyEmpty has an interruptible readMVar operation inside. So if dupRef can't fail, then isStructurallyEmpty can

good point!

Reply:
The existing code is actually ok, but I'll refactor it slightly and add a comment to clarify why:
```
    mts' <- V.filterM (fmap not . isStructurallyEmpty) (V.fromList mts)
    -- isStructurallyEmpty could receive async exceptions even when masked,
    -- but we use it before allocating new references.
    mts'' <- V.mapM dupRef mts'
```

Reply:
And actually `newPendingLevelMerge` did have the problem, since there we allocated first and then used interruptible `readMVar`. I've swapped it around and clarified the comments.